dhis-2/dhis-e2e-test/src/test/java/org/hisp/dhis/tracker/importer/relationships/RelationshipsTests.java

```diff
@@ -165,6 +170,102 @@ public void shouldImportObjectsWithRelationship( String file )
+        IdGenerator idGenerator = new IdGenerator();
+
+        String trackedEntity_1 = idGenerator.generateUniqueId();
+        String trackedEntity_2 = idGenerator.generateUniqueId();
+
+        JsonObject jsonObject = new JsonObjectBuilder()
+            .addArray( "trackedEntities",
+                buildTrackedEntity( trackedEntity_1 ),
+                buildTrackedEntity( trackedEntity_2 ) )
+            .addArray( "relationships",
+                buildNonBidirectionalRelationship( trackedEntity_1, trackedEntity_2 ) )
+            .build();
+
+        TrackerApiResponse trackerApiResponse = trackerActions.postAndGetJobReport( jsonObject )
+            .validateSuccessfulImport();
+
+        trackerApiResponse
+            .validate()
+            .body( "bundleReport.typeReportMap.RELATIONSHIP.stats.created", equalTo( 1 ) );
+
+        String createdRelationshipUid = trackerApiResponse.extractImportedRelationships().get( 0 );
+
+        // when posting the same payload, then relationship is ignored when in
+        // the same way
+        trackerActions.postAndGetJobReport( jsonObject )
+            .validateSuccessfulImport()
+            .validate()
+            .body( "bundleReport.typeReportMap.RELATIONSHIP.stats.ignored", equalTo( 1 ) );
+
+        // and is imported again if the relation is in inverse order
+        trackerActions.postAndGetJobReport( invertRelationship( jsonObject ) )
+            .validateSuccessfulImport()
+            .validate()
+            .body( "bundleReport.typeReportMap.RELATIONSHIP.stats.ignored", equalTo( 0 ) )
+            .body( "bundleReport.typeReportMap.RELATIONSHIP.stats.created", equalTo( 1 ) );
+
+        // and relationship is not duplicated
+        ApiResponse relationshipResponse = trackerActions.get( "/relationships?tei=" + trackedEntity_1 );
+
+        relationshipResponse
+            .validate()
+            .statusCode( 200 )
+            .body( "size()", is( 2 ) );
+    }
+
+    @Test
+    public void shouldNotDuplicateBidirectionalRelationship()
+    {
+        // Given 2 existing tracked entities and a bidirectional relationship
+        // between them
+        IdGenerator idGenerator = new IdGenerator();
+
+        String trackedEntity_1 = idGenerator.generateUniqueId();
+        String trackedEntity_2 = idGenerator.generateUniqueId();
+
+        JsonObject jsonObject = new JsonObjectBuilder()
+            .addArray( "trackedEntities",
+                buildTrackedEntity( trackedEntity_1 ),
+                buildTrackedEntity( trackedEntity_2 ) )
+            .addArray( "relationships",
+                buildBidirectionalRelationship( trackedEntity_1, trackedEntity_2 ) )
+            .build();
+
+        TrackerApiResponse trackerApiResponse = trackerActions.postAndGetJobReport( jsonObject )
+            .validateSuccessfulImport();
+
+        trackerApiResponse
+            .validate()
+            .body( "bundleReport.typeReportMap.RELATIONSHIP.stats.created", equalTo( 1 ) );
```

Comment:
I would suggest to create a new method in TrackerApiResponse, similar to validateEvents(), validateEnrollments(). It would be easier to update if we decide to change the response again

Reply:
agree